pkg/apis/resource/validation/validation.go

```diff
@@ -17,172 +17,264 @@ limitations under the License.
 package validation
 
 import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"regexp"
+	"strings"
+
 	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	apiresource "k8s.io/apimachinery/pkg/api/resource"
 	apimachineryvalidation "k8s.io/apimachinery/pkg/api/validation"
-	pathvalidation "k8s.io/apimachinery/pkg/api/validation/path"
 	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/apimachinery/pkg/util/sets"
-	utilvalidation "k8s.io/apimachinery/pkg/util/validation"
+	"k8s.io/apimachinery/pkg/util/validation"
 	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/apiserver/pkg/cel"
+	"k8s.io/apiserver/pkg/cel/environment"
+	dracel "k8s.io/dynamic-resource-allocation/cel"
 	corevalidation "k8s.io/kubernetes/pkg/apis/core/validation"
 	"k8s.io/kubernetes/pkg/apis/resource"
-	namedresourcesvalidation "k8s.io/kubernetes/pkg/apis/resource/structured/namedresources/validation"
 )
 
-// validateResourceDriverName reuses the validation of a CSI driver because
-// the allowed values are exactly the same.
-var validateResourceDriverName = corevalidation.ValidateCSIDriverName
+var (
+	// validateResourceDriverName reuses the validation of a CSI driver because
+	// the allowed values are exactly the same.
+	validateDriverName      = corevalidation.ValidateCSIDriverName
+	validateDeviceName      = corevalidation.ValidateDNS1123Subdomain
+	validateDeviceClassName = corevalidation.ValidateDNS1123Subdomain
+	validateRequestName     = corevalidation.ValidateDNS1123Label
+)
 
-// ValidateClaim validates a ResourceClaim.
-func ValidateClaim(resourceClaim *resource.ResourceClaim) field.ErrorList {
-	allErrs := corevalidation.ValidateObjectMeta(&resourceClaim.ObjectMeta, true, corevalidation.ValidateResourceClaimName, field.NewPath("metadata"))
-	allErrs = append(allErrs, validateResourceClaimSpec(&resourceClaim.Spec, field.NewPath("spec"))...)
+func validatePoolName(name string, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if name == "" {
+		allErrs = append(allErrs, field.Required(fldPath, ""))
+	} else {
+		if len(name) > resource.PoolNameMaxLength {
+			allErrs = append(allErrs, field.TooLongMaxLength(fldPath, name, resource.PoolNameMaxLength))
+		}
+		parts := strings.Split(name, "/")
+		for i, part := range parts {
+			idxPath := fldPath
+			if len(parts) > 1 {
+				idxPath = idxPath.Index(i)
```

Comment:
I don't think copying ValidateDNS1123Subdomain just to get a better message is worth it, in particular because this is for driver developers, not end users. I've removed the indexing.